### ISSUE TITLE
Impossible conditionals should be errors

### ISSUE THREAD

**Issue by wincent** [microsoft/TypeScript#37804]
## Search Terms

- "unreachable"

Possibly duplicates https://github.com/microsoft/TypeScript/issues/34955 (but I am not sure).

## Suggestion

I'd like to see TypeScript complain about (most) conditionals that can never possibly be true. [Example](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAHjAvDA2gRgLoG4BQOCWAZjABQL6RQCGYwApiMQMpQBOFA5gJQwDeOMQTFCQQAGzoA6MSA4kA5AFcwrOlWAALKgCMJ8rjgC+QA):

```ts
const x = [1];

if (x instanceof String) {
    // Should complain here...
    console.log('unreachable')
}
```

This is a contrived toy example; see below for a motivating use case.

Given that this increases the range of conditions under which errors may be newly emitted, I don't know whether this would need to be gated behind a compiler option that people desiring more strictness could opt-in to.

## Use Cases

In [real world code](https://www.typescriptlang.org/play?#code/PTAEBUAsEsGdQGYFcB2BjALtA9i0ATbAU3lmwFsiMYUBzUAQ1gE90AaUAG2gGsjRyDHtDqNQKKgHdsAJx6gZRAI5ISGAHQAofETScGixKkw48CbNgAUASgBcoAAoyKcIgB4AIgwz8APuKROTgA+AG5NTSZWNCN0LFxQNAYgohkbUABvTVAcxNxYDFAAN2TQAF5ECxtw7NzoBFBLEs5QEQKGdCJsBq8fa0za3Ny0fOxOInVObFpLAHJURQY0SAYAI3GOVd0GJFh+SX5zGVpsQoZJBmgMWetB0ABfTXugA), a common mistake is to forget to `await` an `async` function:

```ts
// This function does something async, like making a network request.
declare function foo(): Promise<Date | null>;

async function caller() {
    // BUG: missing await keyword
    const val = foo();

    if (val instanceof Date) {
        console.log('unreachable, because we forgot await')
    }
}
```

## Examples

Ideally, TS would complain that the `if (val instanceof Date)` conditional is impossible and that the `console.log` is therefore unreachable and very likely indicative of a programming error.

Counterexamples which are not likely to be indicative of programming errors are things like then following, which are most likely to intentional/temporary deactivations of code:

```ts
if (false) {
  doSomething();
}

if (0) {
  doSomethingElse();
}
```

I'd suggest that for these and perhaps a small range of other cases TS should refrain from complaining.

## Checklist

Suggestion meets these guidelines? Not sure, but I think I can tick off at least a couple of them:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by jack-williams** [comment]
I think this is probably a duplicate of #32801 + #30646.

TypeScript already produces errors for some set of conditions that are always known to be false. The proposal in #32801 would add your example to the list of errors produced. Once you have an error, it might be reasonable to expect a quickfix that suggests inserting `await` to make the condition non-trivial.
